File: src/contract/targets.mjs
/**
 * Reference implementation of the Scrollcase box-format target model.
 *
 * A target is the (platform, arch, accelerator) triple a box is built for, plus a CUDA ABI version
 * when the accelerator is CUDA. `boxTargetId()` turns it into the canonical slug that appears
 * in archive names, object keys, and registry routes, so every implementation of the format — this
 * one, a consumer's own client, a signer — must agree character for character. The golden cases in
 * `fixtures/target-id-contract.json` are what "agree" means, and are the fixtures other languages
 * validate their mirrors against.
 *
 * The adapters below describe what a target implies for the built payload: the conda subdir it
 * solves for, the archive backend, how native libraries are inspected, and the environment a
 * validation run gets. They are part of the format because a consumer unpacking a box relies on
 * them.
 *
 * What a target deliberately no longer describes is the *runtime* inside the box. The interpreter
 * layout, the execution kinds and the runtime's own environment variables live in `runtimes.mjs`,
 * because they are facts about what a box runs rather than about the machine it runs on. Keeping
 * them here made every target adapter a statement that a box is a Python box, and made a second
 * runtime a fork of this table rather than one more adapter beside it.
 */
/**
 * What a target implies for the built payload. Part of the format rather than an implementation
 * detail: a consumer unpacking a box relies on this.
 *
 * @typedef {object} BoxTargetAdapter
 * @property {string} id canonical adapter id, e.g. `macos-aarch64`
 * @property {'macos' | 'linux' | 'windows'} platform
 * @property {'aarch64' | 'x86_64'} arch
 * @property {{ platform: string, arch: string }} host the Node platform/arch a build must run on
 * @property {'osx-arm64' | 'linux-64' | 'win-64'} condaSubdir the scroll's pixi `platforms` value
 * @property {{ format: 'zip', writer: string, reader: string, assetTarReader: string,
 *   zip64: boolean }} archive the pinned archive backend
 * @property {{ command: string, argsPrefix: readonly string[],
 *   extensions: readonly string[] }} nativeLibraryInspection
 * @property {Readonly<Record<string, Readonly<Record<string, string>>>>} validationEnvironments
 *   the environment that forces a run onto one accelerator, keyed by accelerator
 * @property {readonly string[]} executionAffectingEnvironmentVariables the operating system's own
 *   dynamic-linker controls; the runtime adds the variables its loader reads, and
 *   `executionAffectingVariables()` in `runtimes.mjs` is what joins the two halves
 */

import { IMPLICIT_RUNTIME_ID, assertRuntimeEntryPoint } from './runtimes.mjs';

const TARGET_ACCELERATORS = {
  macos: { aarch64: ['metal', 'cpu'] },
  linux: { x86_64: ['cpu', 'cuda'] },
  windows: { x86_64: ['cpu', 'cuda'] },
};
const CUDA_VERSION = /^[1-9][0-9]*\.[0-9]+$/;

// The exact libraries that wrote and read a box, so a consumer knows what produced the bytes it
// holds rather than inferring it. Each version is the one this package installs: they are pinned in
// `package.json` and `tests/unit/contract-targets.test.mjs` fails when the two drift, because a
// descriptor naming a release that never touched the archive is worse than no descriptor at all.
const ARCHIVE_BACKEND = Object.freeze({
  format: 'zip',
  writer: 'yazl@3.3.1',
  reader: 'yauzl@3.4.0',
  assetTarReader: 'tar@7.5.22',
  zip64: true,
});

const TARGET_ADAPTERS = Object.freeze([
  Object.freeze({
    id: 'macos-aarch64',
    platform: 'macos',
    arch: 'aarch64',
    host: Object.freeze({ platform: 'darwin', arch: 'arm64' }),
    // conda platform subdir: the `platforms` value in the scroll's pixi.toml.
    condaSubdir: 'osx-arm64',
    archive: ARCHIVE_BACKEND,
    nativeLibraryInspection: Object.freeze({
      command: 'otool',
      argsPrefix: Object.freeze(['-L']),
      extensions: Object.freeze(['.dylib', '.so']),
    }),
    validationEnvironments: Object.freeze({
      cpu: Object.freeze({ CUDA_VISIBLE_DEVICES: '' }),
      metal: Object.freeze({ PYTORCH_ENABLE_MPS_FALLBACK: '0' }),
    }),
    executionAffectingEnvironmentVariables: Object.freeze(['DYLD_INSERT_LIBRARIES']),
  }),
  Object.freeze({
    id: 'linux-x86_64',
    platform: 'linux',
    arch: 'x86_64',
    host: Object.freeze({ platform: 'linux', arch: 'x64' }),
    condaSubdir: 'linux-64',
    archive: ARCHIVE_BACKEND,
    nativeLibraryInspection: Object.freeze({
      command: 'ldd',
      argsPrefix: Object.freeze([]),
      extensions: Object.freeze(['.so']),
    }),
    validationEnvironments: Object.freeze({
      cpu: Object.freeze({ CUDA_VISIBLE_DEVICES: '' }),
      cuda: Object.freeze({ CUDA_VISIBLE_DEVICES: '0' }),
    }),
    executionAffectingEnvironmentVariables: Object.freeze(['LD_PRELOAD']),
  }),
  Object.freeze({
    id: 'windows-x86_64',
    platform: 'windows',
    arch: 'x86_64',
    host: Object.freeze({ platform: 'win32', arch: 'x64' }),
    condaSubdir: 'win-64',
    archive: ARCHIVE_BACKEND,
    nativeLibraryInspection: Object.freeze({
      command: 'dumpbin',
      argsPrefix: Object.freeze(['/DEPENDENTS']),
      extensions: Object.freeze(['.dll', '.pyd']),
    }),
    validationEnvironments: Object.freeze({
      cpu: Object.freeze({ CUDA_VISIBLE_DEVICES: '' }),
      cuda: Object.freeze({ CUDA_VISIBLE_DEVICES: '0' }),
    }),
    // Windows has no inherited loader control of its own worth reporting: `PATH` decides DLL
    // resolution and is far too broad to name here, so the whole list is the runtime's.
    executionAffectingEnvironmentVariables: Object.freeze([]),
  }),
]);

/**
 * Returns the canonical target slug used in box filenames, object keys, and routes.
 *
 * @param {import('./types/index.d.ts').BoxTarget} target
 * @returns {string} the canonical slug, e.g. `linux-x86_64-cuda12.4`
 * @throws {TypeError} when the target is outside the supported matrix, or its CUDA version is
 *   missing on a CUDA target or present on any other
 */
export function boxTargetId(target) {
  if (!target || typeof target !== 'object') {
    throw new TypeError('Box target must be an object');
  }
  const accelerators = TARGET_ACCELERATORS[target?.platform]?.[target?.arch];
  if (!accelerators?.includes(target?.accelerator)) {
    throw new TypeError(
      `Unsupported box target: ${target?.platform}/${target?.arch}/${target?.accelerator}`,
    );
  }
  if (target.accelerator === 'cuda') {
    if (typeof target.cudaVersion !== 'string' || !CUDA_VERSION.test(target.cudaVersion)) {
      throw new TypeError('A CUDA box target requires a numeric major.minor CUDA version');
    }
    return `${target.platform}-${target.arch}-cuda${target.cudaVersion}`;
  }
  if (target.cudaVersion !== undefined) {
    throw new TypeError('Only CUDA box targets may declare a CUDA version');
  }
  return `${target.platform}-${target.arch}-${target.accelerator}`;
}

/**
 * Returns the native builder adapter for a validated box target.
 *
 * @param {import('./types/index.d.ts').BoxTarget} target
 * @returns {BoxTargetAdapter}
 * @throws {TypeError} when the target is unsupported
 */
export function boxTargetAdapter(target) {
  boxTargetId(target);
  const adapter = TARGET_ADAPTERS.find((candidate) =>
    candidate.platform === target.platform && candidate.arch === target.arch);
  if (!adapter) throw new TypeError(`No box target adapter exists for ${target.platform}/${target.arch}`);
  return adapter;
}

/**
 * Ensures a build or target lock runs on the OS and architecture it will ship for.
 *
 * @param {BoxTargetAdapter} adapter
 * @param {{ platform: string, arch: string }} [host] defaults to the current process
 * @returns {void}
 * @throws {TypeError} when the host is not the OS and architecture the box ships for
 */
export function assertNativeHost(adapter, host = process) {
  if (host.platform !== adapter.host.platform || host.arch !== adapter.host.arch) {
    throw new TypeError(
      `${adapter.id} boxes must be built natively on ${adapter.host.platform}/${adapter.host.arch}; `
      + `current host is ${host.platform}/${host.arch}`,
    );
  }
}

/**
 * Ensures the scroll entry point agrees with the standalone Python layout for this target.
 *
 * Kept under its published name while the wire format still spells the field `pythonEntryPoint`.
 * The rule itself moved to `runtimes.mjs`, where it can be asked about any runtime; this is the one
 * public spelling of it, and it goes when the field does.
 *
 * @param {BoxTargetAdapter} adapter
 * @param {string} entryPoint
 * @returns {void}
 * @throws {TypeError} when the entry point does not match the runtime's layout for this target
 */
export function assertPythonEntryPoint(adapter, entryPoint) {
  assertRuntimeEntryPoint(IMPLICIT_RUNTIME_ID, adapter, entryPoint);
}

/**
 * Lists every adapter, for contract tests and for consumers enumerating supported targets.
 *
 * @returns {BoxTargetAdapter[]} every supported adapter, as a fresh array
 */
export function boxTargetAdapters() {
  return [...TARGET_ADAPTERS];
}

/**
 * Maps a validated box target to its conda platform subdir (the pixi `platforms` value).
 *
 * @param {import('./types/index.d.ts').BoxTarget} target
 * @returns {'osx-arm64' | 'linux-64' | 'win-64'} the pixi `platforms` value for the target
 */
export function condaSubdir(target) {
  const adapter = boxTargetAdapter(target);
  return adapter.condaSubdir;
}

/**
 * Returns the conda accelerator descriptor a scroll selects, rejecting target drift. `metal` and
 * `cpu` need no extra conda knobs (osx-arm64 ships MPS in the pytorch build; cpu is the default build); `cuda` pins a
 * `cuda-version` and declares a CUDA system requirement so the solver picks the GPU pytorch build.
 *
 * @param {Pick<import('./types/index.d.ts').BoxScroll, 'target'>} scroll
 * @returns {{ accelerator: 'cpu' | 'metal' | 'cuda', cudaVersion: string | null }}
 * @throws {TypeError} when the accelerator is unsupported, or a CUDA target lacks a version
 */
export function pixiAccelerator(scroll) {
  const accelerator = scroll?.target?.accelerator;
  if (accelerator === 'metal' || accelerator === 'cpu') {
    return Object.freeze({ accelerator, cudaVersion: null });
  }
  if (accelerator === 'cuda') {
    const cudaVersion = scroll?.target?.cudaVersion;
    if (typeof cudaVersion !== 'string' || !CUDA_VERSION.test(cudaVersion)) {
      throw new TypeError('A CUDA box target requires a numeric major.minor CUDA version');
    }
    return Object.freeze({ accelerator, cudaVersion });
  }
  throw new TypeError(`Unsupported box accelerator: ${accelerator}`);
}

File: rust/src/contract/targets.rs
//! Mirror of the Scrollcase box-format target model.
//!
//! A target is the `(platform, arch, accelerator)` triple a box is built for, plus a CUDA ABI
//! version when the accelerator is CUDA. [`box_target_id`] turns it into the canonical slug that
//! appears in archive names, object keys and registry routes, so every implementation of the format
//! must agree character for character. The golden cases in `fixtures/target-id-contract.json` are
//! what "agree" means, and `tests/contract.rs` proves this mirror against them.
//!
//! The adapter describes what a target implies for the extracted tree. Only the parts a consumer
//! relies on are carried here: the native host it may run on, and the operating system's own
//! dynamic-linker controls. The builder's own adapter additionally names the archive backend, the
//! conda subdir and the native library inspector — all of them decisions taken while a box is
//! produced, none of them observable by something that only unpacks and runs one.
//!
//! What a target deliberately no longer describes is the *runtime* inside the box. The interpreter
//! layout, the execution kinds and the runtime's own environment variables live in
//! [`super::runtimes`], because they are facts about what a box runs rather than about the machine
//! it runs on.
//!
//! The native host is expressed in Rust's own `OS`/`ARCH` vocabulary rather than Node's
//! `darwin`/`arm64`. Those strings never appear in a signed document; they only answer "may this
//! host run this box", and each implementation answers it in the terms its own runtime reports.

use serde::{Deserialize, Serialize};

use crate::error::{fail, Result};

/// The `(platform, arch, accelerator)` triple a box is built for.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BoxTarget {
    /// Operating system the box runs on.
    pub platform: String,
    /// CPU architecture the box runs on.
    pub arch: String,
    /// Compute backend the box was built against.
    pub accelerator: String,
    /// CUDA ABI version, required on a CUDA target and forbidden on every other.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cuda_version: Option<String>,
}

/// What a target implies for the extracted tree.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BoxTargetAdapter {
    /// Canonical adapter id, for example `macos-aarch64`.
    pub id: &'static str,
    /// Operating system, as a scroll declares it.
    pub platform: &'static str,
    /// CPU architecture, as a scroll declares it.
    pub arch: &'static str,
    /// `std::env::consts::OS` value a host must report to run this box.
    pub host_os: &'static str,
    /// `std::env::consts::ARCH` value a host must report to run this box.
    pub host_arch: &'static str,
    /// The operating system's own dynamic-linker controls; the runtime adds the variables its
    /// loader reads, and [`super::runtimes::execution_affecting_variables`] joins the two halves.
    pub execution_affecting_environment_variables: &'static [&'static str],
}

const MACOS_EXECUTION_ENVIRONMENT: &[&str] = &["DYLD_INSERT_LIBRARIES"];

const LINUX_EXECUTION_ENVIRONMENT: &[&str] = &["LD_PRELOAD"];

// Windows has no inherited loader control of its own worth reporting: `PATH` decides DLL
// resolution and is far too broad to name here, so the whole list is the runtime's.
const WINDOWS_EXECUTION_ENVIRONMENT: &[&str] = &[];

const TARGET_ADAPTERS: &[BoxTargetAdapter] = &[
    BoxTargetAdapter {
        id: "macos-aarch64",
        platform: "macos",
        arch: "aarch64",
        host_os: "macos",
        host_arch: "aarch64",
        execution_affecting_environment_variables: MACOS_EXECUTION_ENVIRONMENT,
    },
    BoxTargetAdapter {
        id: "linux-x86_64",
        platform: "linux",
        arch: "x86_64",
        host_os: "linux",
        host_arch: "x86_64",
        execution_affecting_environment_variables: LINUX_EXECUTION_ENVIRONMENT,
    },
    BoxTargetAdapter {
        id: "windows-x86_64",
        platform: "windows",
        arch: "x86_64",
        host_os: "windows",
        host_arch: "x86_64",
        execution_affecting_environment_variables: WINDOWS_EXECUTION_ENVIRONMENT,
    },
];

/// The accelerators each `(platform, arch)` pair supports, in the order the format defines them.
fn supported_accelerators(platform: &str, arch: &str) -> Option<&'static [&'static str]> {
    match (platform, arch) {
        ("macos", "aarch64") => Some(&["metal", "cpu"]),
        ("linux" | "windows", "x86_64") => Some(&["cpu", "cuda"]),
        _ => None,
    }
}

/// Whether a CUDA version is the `major.minor` shape the format requires.
///
/// Hand-written rather than delegated to a regex engine: the pattern is
/// `^[1-9][0-9]*\.[0-9]+$`, and carrying a regex dependency to answer it would be the whole cost of
/// the crate's smallest question.
fn is_cuda_version(value: &str) -> bool {
    let Some((major, minor)) = value.split_once('.') else {
        return false;
    };
    let major_valid = !major.is_empty()
        && !major.starts_with('0')
        && major.bytes().all(|byte| byte.is_ascii_digit());
    let minor_valid = !minor.is_empty() && minor.bytes().all(|byte| byte.is_ascii_digit());
    major_valid && minor_valid
}

/// Returns the canonical target slug used in box filenames, object keys and routes.
///
/// # Errors
///
/// When the target is outside the supported matrix, or its CUDA version is missing on a CUDA target
/// or present on any other.
pub fn box_target_id(target: &BoxTarget) -> Result<String> {
    let accelerators = supported_accelerators(&target.platform, &target.arch);
    if !accelerators.is_some_and(|values| values.contains(&target.accelerator.as_str())) {
        fail!(
            "Unsupported box target: {}/{}/{}",
            target.platform,
            target.arch,
            target.accelerator
        );
    }
    if target.accelerator == "cuda" {
        let Some(version) = target.cuda_version.as_deref().filter(|v| is_cuda_version(v)) else {
            fail!("A CUDA box target requires a numeric major.minor CUDA version");
        };
        return Ok(format!(
            "{}-{}-cuda{version}",
            target.platform, target.arch
        ));
    }
    if target.cuda_version.is_some() {
        fail!("Only CUDA box targets may declare a CUDA version");
    }
    Ok(format!(
        "{}-{}-{}",
        target.platform, target.arch, target.accelerator
    ))
}

/// Returns the adapter for a validated box target.
///
/// # Errors
///
/// When the target is unsupported.
pub fn box_target_adapter(target: &BoxTarget) -> Result<&'static BoxTargetAdapter> {
    box_target_id(target)?;
    let Some(adapter) = TARGET_ADAPTERS
        .iter()
        .find(|candidate| candidate.platform == target.platform && candidate.arch == target.arch)
    else {
        fail!(
            "No box target adapter exists for {}/{}",
            target.platform,
            target.arch
        );
    };
    Ok(adapter)
}

/// Lists every adapter, for contract tests and for callers enumerating supported targets.
#[must_use]
pub fn box_target_adapters() -> &'static [BoxTargetAdapter] {
    TARGET_ADAPTERS
}

/// Ensures this host is the operating system and architecture the box ships for.
///
/// # Errors
///
/// When the current host is not the one the adapter requires.
pub fn assert_native_host(adapter: &BoxTargetAdapter) -> Result<()> {
    assert_host(adapter, std::env::consts::OS, std::env::consts::ARCH)
}

/// The host check with the host injected, so every target can be exercised on one machine.
///
/// # Errors
///
/// When the supplied host is not the one the adapter requires.
pub fn assert_host(adapter: &BoxTargetAdapter, os: &str, arch: &str) -> Result<()> {
    if os != adapter.host_os || arch != adapter.host_arch {
        fail!(
            "{} boxes cannot run on {os}/{arch}; they require {}/{}",
            adapter.id,
            adapter.host_os,
            adapter.host_arch
        );
    }
    Ok(())
}

/// Ensures a release's entry point agrees with the standalone Python layout for this target.
///
/// Kept under its published name while the wire format still spells the field `pythonEntryPoint`.
/// The rule itself lives in [`super::runtimes`], where it can be asked about any runtime.
///
/// # Errors
///
/// When the entry point is not the one the runtime defines for this target.
pub fn assert_python_entry_point(adapter: &BoxTargetAdapter, entry_point: &str) -> Result<()> {
    super::runtimes::assert_runtime_entry_point(
        super::runtimes::IMPLICIT_RUNTIME_ID,
        adapter,
        entry_point,
    )
}

#[cfg(test)]
mod tests {
    use super::{
        assert_host, assert_python_entry_point, box_target_adapter, box_target_adapters,
        box_target_id, is_cuda_version, BoxTarget,
    };

    fn target(platform: &str, arch: &str, accelerator: &str, cuda: Option<&str>) -> BoxTarget {
        BoxTarget {
            platform: platform.to_string(),
            arch: arch.to_string(),
            accelerator: accelerator.to_string(),
            cuda_version: cuda.map(str::to_string),
        }
    }

    #[test]
    fn cuda_versions_follow_the_major_minor_rule() {
        assert!(is_cuda_version("12.4"));
        assert!(is_cuda_version("9.0"));
        for invalid in ["", "12", "12.", ".4", "0.1", "01.2", "12.4.1", "12.x", " 12.4"] {
            assert!(!is_cuda_version(invalid), "{invalid} was accepted");
        }
    }

    #[test]
    fn every_adapter_is_reachable_from_a_target() {
        for adapter in box_target_adapters() {
            let accelerator = if adapter.platform == "macos" {
                "metal"
            } else {
                "cpu"
            };
            let resolved =
                box_target_adapter(&target(adapter.platform, adapter.arch, accelerator, None))
                    .unwrap();
            assert_eq!(resolved.id, adapter.id);
        }
    }

    #[test]
    fn a_foreign_host_is_refused_with_the_shared_wording() {
        let adapter = box_target_adapter(&target("linux", "x86_64", "cpu", None)).unwrap();
        assert!(assert_host(adapter, "linux", "x86_64").is_ok());
        let error = assert_host(adapter, "macos", "aarch64").unwrap_err();
        assert!(error.message().contains("cannot run on"), "{error}");
    }

    #[test]
    fn an_entry_point_from_another_platform_is_refused() {
        let windows = box_target_adapter(&target("windows", "x86_64", "cpu", None)).unwrap();
        assert!(assert_python_entry_point(windows, "venv/python.exe").is_ok());
        assert!(assert_python_entry_point(windows, "venv/bin/python").is_err());
    }

    #[test]
    fn a_target_id_is_never_produced_for_an_unsupported_triple() {
        assert!(box_target_id(&target("macos", "x86_64", "cpu", None)).is_err());
        assert!(box_target_id(&target("linux", "x86_64", "metal", None)).is_err());
    }
}
